cardano-api/src/Cardano/Api/TxBody.hs

```diff
@@ -216,6 +227,75 @@ import           Cardano.Ledger.Crypto (StandardCrypto)
 {- HLINT ignore "Redundant flip" -}
 {- HLINT ignore "Use section" -}
 
+-- | Indicates whether a script is expected to fail or pass validation.
+data ScriptValidity
+  = ScriptInvalid -- ^ Script is expected to fail validation.
+                  -- Transactions marked as such can include scripts that fail validation.
+                  -- Such transactions may be submitted to the chain, in which case the
+                  -- collateral will be taken upon on chain script validation failure.
+
+  | ScriptValid   -- ^ Script is expected to pass validation.
+                  -- Transactions marked as such cannot include scripts that fail validation.
+
+  deriving (Eq, Show)
+
+instance ToCBOR ScriptValidity where
+  toCBOR = toCBOR . scriptValidityToIsValid
+
+instance FromCBOR ScriptValidity where
+  fromCBOR = isValidToScriptValidity <$> fromCBOR
+
+scriptValidityToIsValid :: ScriptValidity -> Alonzo.IsValid
+scriptValidityToIsValid ScriptInvalid = Alonzo.IsValid False
+scriptValidityToIsValid ScriptValid = Alonzo.IsValid True
+
+isValidToScriptValidity :: Alonzo.IsValid -> ScriptValidity
+isValidToScriptValidity (Alonzo.IsValid False) = ScriptInvalid
+isValidToScriptValidity (Alonzo.IsValid True) = ScriptValid
+
+-- | A representation of whether the era supports tx script validity.
+--
+-- The Mary and subsequent eras support script validity.
+--
+data TxBodyScriptValidity era where
+  TxBodyScriptValidityNone :: TxBodyScriptValidity era
+
+  -- | Tx script validity is supported in transactions in the 'Alonzo' era.
+  TxBodyScriptValidity
+    :: TxBodyScriptValiditySupportedInEra era
+    -> ScriptValidity
+    -> TxBodyScriptValidity era
+
+deriving instance Eq   (TxBodyScriptValiditySupportedInEra era)
+deriving instance Show (TxBodyScriptValiditySupportedInEra era)
+
+data TxBodyScriptValiditySupportedInEra era where
+  TxBodyScriptValiditySupportedInAlonzoEra :: TxBodyScriptValiditySupportedInEra AlonzoEra
+
+deriving instance Eq   (TxBodyScriptValidity era)
+deriving instance Show (TxBodyScriptValidity era)
+
+txBodyScriptValiditySupportedInEra :: CardanoEra era -> Maybe (TxBodyScriptValiditySupportedInEra era)
+txBodyScriptValiditySupportedInEra ByronEra   = Nothing
+txBodyScriptValiditySupportedInEra ShelleyEra = Nothing
+txBodyScriptValiditySupportedInEra AllegraEra = Nothing
+txBodyScriptValiditySupportedInEra MaryEra    = Nothing
+txBodyScriptValiditySupportedInEra AlonzoEra  = Just TxBodyScriptValiditySupportedInAlonzoEra
+
+txBodyScriptValidityToScriptValidity :: TxBodyScriptValidity era -> ScriptValidity
```

Comment:
This should be parameterised with `ShelleyBasedEra era` so we can utilize the `EmptyCase` extension to disallow the Byron era.

Reply:
If we do this, we can no longer implement:

```
makeTransactionBody :: forall era.
     IsCardanoEra era
  => TxBodyScriptValidity era -- ^ Mark script as expected to pass or fail validation
  -> TxBodyContent BuildTx era
  -> Either TxBodyError (TxBody era)
```

Which must accept `TxBodyScriptValidity era`.  Otherwise there will be no means to thread through the script validity.

Reply:
Modifying `TxBodyContent` to include script validity solve this problem for me.

Reply:
I made the change and discovered I actually need the `CardanoEra era` argument in some cases.  For example the generators and the CLI parsers.

So I created a separate function for the `ShelleyBasedEra era` case.  Turns out the use case for `ShelleyBasedEra era` are all in the same module.